src/pages/LoginPage/login-page.jsx

```diff
@@ -0,0 +1,89 @@
+/* eslint-disable no-undef */
+import {
+  Button,
+  EmailInput,
+  PasswordInput
+} from '@ya.praktikum/react-developer-burger-ui-components'
+import { useCallback, useState } from 'react'
+import { useDispatch } from 'react-redux'
+import { Link } from 'react-router-dom'
+import { getUser } from '../../services/actions/user'
+import style from './login-page.module.css'
+import { Redirect, useLocation } from 'react-router-dom'
+
+export const LoginPage = ({ userReceived, isAutorized }) => {
+  const location = useLocation()
+  const [form, setValue] = useState({ email: '', password: '' })
+  const dispatch = useDispatch()
+
+  const onChange = e => {
+    setValue({ ...form, [e.target.name]: e.target.value })
+  }
+
+  const login = useCallback(
+    async e => {
+      e.preventDefault()
+      await dispatch(getUser(form))
+    },
+    [dispatch, form]
+  )
+
+  if (!userReceived) {
+    return null
+  }
+
+  if (isAutorized) {
+    return <Redirect to={location.state?.from || '/'} />
```

Comment:
Надо исправить: Мне кажется, что вместо того, чтобы оборачивать компоненты HOC. И в каждом писать `if (isAutorized) { return <Redirect to={location.state?.from || '/'} />` стоит вынести этот функционал на `protectedRote`. К примеру:
Можно добавить для `ProtectedRote` пропс `onlyAuth`, если он `ture`, то он пропускает только авторизированных пользователей, иначе только не авторизированных. Так мы сможем вынести весь функционал редиректов в 1 компонент и не будем думать о том, как это делать в самом компоненте.

Reply:
При таком сценарии приходится оборачивать практически все роуты в ProtectedRoute:

<Switch location={background || location}>
          <ProtectedRoute path="/login" isOnlyAuth={true} exact={true}>
            <LoginPage />
          </ProtectedRoute>

          <ProtectedRoute path="/register" isOnlyAuth={true} exact={true}>
            <RegisterPage />
          </ProtectedRoute>

          <ProtectedRoute
            path="/forgot-password"
            isOnlyAuth={true}
            exact={true}
          >
            <ForgotPassPage />
          </ProtectedRoute>

          <ProtectedRoute path="/reset-password" isOnlyAuth={true} exact={true}>
            <ResetPassPage />
          </ProtectedRoute>
         ...
</Swicth>

Это приводит к ситуации, когда для каждого ProtectedRoute вызывается его внутренняя логика, внутри которой есть свой state:

const [userReceived, setUserReceived] = useState(false)

который нужен для того, чтобы компонент не рендерил элементы до момента, пока не закончится асинхронный вызов функции getUserWithToken. 
Проблема в том, что при вызове на каждый объявленный в App, ProtectedRoute изменение состояния приводило к ошибке:

Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops

Поэтому данный подход не работает, но могу заблуждаться, просьба помочь разобраться, если подход с HOC не приемлем.

Reply:
Такс. Давайте разбираться:

`Это приводит к ситуации, когда для каждого ProtectedRoute вызывается его внутренняя логика, внутри которой есть свой state`
- Вы все равно создаете `state` внутри HOC и ждем ответа сервера. Так что в плане перфоманса тут никакой разницы нет. 
Что про ошибку, то я бы использовал вместо state состояние в redux. Вроде `isAuthChecked` а в app просто бы вызывал проверку, которая бы устанавливала это состояние в `true` или `false`
Тогда код `ProtectedRoute` бы был примерно такой:

```jsx
const isAuthChecked = useSelector((state) => state.user.isAuthChecked);
  const user = useSelector((state) => state.user.data);
  const location = useLocation();

  if (!isAuthChecked) {
    return <Preloader />;
  }

  if (onlyUnAuth && user) {
    const { from } = location.state || { from: { pathname: "/" } };
    return <Redirect to={from} />;
  }

  if (!onlyUnAuth && !user) {
    return (
      <Redirect
        to={{
          pathname: "/login",
          state: { from: location },
        }}
      />
    );
  }

  return <Route {...rest} />;

```

По сути то, что у вас есть это почти готовый вариант, и там и там используется HOC компонент (в вашем случае это скорее функция высшего порядка, но это не важно)
Скорее просто код будет более структурированным в данном случае. Не приходится создавать свой компонент к каждой уже имеющися странице. 
А сейчас получается, что вы пишите один и тот же код на каждой странице
![image](https://user-images.githubusercontent.com/51288221/201493770-58b17eaf-c718-4c20-aaf2-4a2ea04d65b4.png)

По сути вы все сделали верно, только необходимо как-то убрать этот код, который к функционалу страницы не имеет отношения, приходится его каждый раз вставлять и логичный вопрос почему бы его не вынести. Вы также можете использовать HOC, только стоит добавить тогда туда логику редиректа. И вместо функции, которая обарачивает компонент. По сути такого хука. Использовать HOC. То есть компонент, который сам по себе ничего не делает, а только расширяет функционал потомков

Reply:
Надеюсь ответил на вопрос. Если что п

Reply:
Спасибо, стало понятнее, но хочу для себя кое чего прояснить. 

Когда react анализирует страницу app он сверяет какие роуты соответствуют url в адресной строке, через обертку switch, соответственно монтируется только один компонент с path, совпадающим с url в браузере, остальные Route и их внутренние компоненты (HOC для страниц регистрации, авторизации, восстановления пароля) не запускаются.

Когда же используются ProtectedRoute, то для каждого при рендере будет выполнятся вся внутренняя логика, т.к. мы по умолчанию заходим во все ProtectedRoute и уже потом внутренний Route сверяет свой path с url в браузерной строке. 

Если я правильно понял жизненный цикл, то все выше описанное говорит о том, что чуть оптимальнее в плане производительности выглядит подход с HOC, но в плане чистоты кода подход с выносом логики HOC для компонентов, доступ к которым есть только у неавторизованных пользователей - не очень.

Reply:
На самом деле react ничего не анализирует. Если грубо смотреть на то как он это делает то это примерно вот так:

```js
Route.jsx

if (window.location == props.to) return props.module
else return null

```
А Switch просто контралирует, чтобы не было 2 наложений экранов друг поверх друга


По сути в данном случае обертка `ProtectedRoute` просто функция. Которая либо возвращает route и тогда все работает как и всегда. Либо редирект (тут вроде понятно), либо просто ситуация вроде Switch и в нем куча прелоадеров. Он в данном случае выбират просто самый 1.

Для точных утверждений нужны коенчно тесты. Но так, на вскидку разницы почти нет

Reply:
Сейчас кстати я видет в каких-то версиях роутера там нет Switch там другой компонет (забыл название)
Вот он правда берет и сравнивает все компоненты по пропсам еще на этапе передачи в компонет-обертку
Но там вообще будет ошибка если мы попытаемся в обертку передать что-то кроме Route